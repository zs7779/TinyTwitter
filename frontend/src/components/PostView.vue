<template>
    <div>
        <post-card :post="post" :verbose=true @like-ok="updatePost($event.id, $event)" @delete-ok="deletePost($event)" v-on="$listeners"></post-card>
    </div>
</template>

<script>
import PostCard from './PostCard.vue'
import { URLs, PLACEHOLDERs } from './utils'

export default{
    name: "user-post-view",
    props: ['username', 'postID'],
    data() {
        return {
            post: PLACEHOLDERs.post,
        }
    },
    methods: {
        getUserPost() {
            axios.get(`${URLs.usersPosts(this.username, this.postID)}`, {
                params: {
                },
            }).then(response => {
                console.log('getPost',response)
                this.post = {
                    ...response.data.post,
                    comments: response.data.comments,
                }
            })
        },
        updatePost(editID, post) {
            if (this.post.id === editID) {
                this.post = post;
            } else this.updateComments(editID, post);
        },
        deletePost(post) {
            if (this.post.id === post.id) this.$router.push({name: 'home'});
            else this.deleteComment(post);
        },
        prependComments(editID, comments) {
            if (this.post.id === editID) {
                comments.forEach(comment => {
                    this.post.comments.unshift(comment);
                    this.post.comment_count = this.post.comments.length;
                });
            }
        },
        appendComments() {},
        updateComments(editID, comment) {
            this.post.comments = this.post.comments.map(p => p.id === editID ? comment : p);    
        },
        addComment(comment) {
            this.prependComments(comment.root_post.id, [comment]);
            this.post.commented = true;
            this.post.comments = this.post.comments.map(p => p.id === comment.parent.id ? {
                ...p,
                comment_count: p.comment_count + 1,
                commented: p.commented + 1,
            } : p);
        },
        deleteComment(comment) {
            this.post.comments = this.post.comments.filter(p => p.id !== comment.id);
            this.post.comment_count = this.post.comments.length;
            this.post.commented = this.post.comments.filter(p => p.owner).length;
            
            this.post.comments = this.post.comments.map(p => p.id === comment.parent.id ? {
                ...p,
                comment_count: p.comment_count - 1,
                commented: p.commented - 1,
            } : p);
        },
        addRepost(post) {
            if (post.parent && post.parent.id == this.post.id) {
                this.post.repost_count++;
                this.post.posted++;
            }
        },
        // delete repost not possible in this view,
    },
    watch: {
        username() {
            this.getUserPost();
        },
        postID() {
            this.getUserPost();
        }
    },
    created() {
        this.getUserPost();
    },
    components: {
        PostCard
    },
}
</script>

<style scoped>
</style>