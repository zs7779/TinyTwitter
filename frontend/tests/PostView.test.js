import { shallowMount } from '@vue/test-utils'
import PostView from '../src/components/PostView'
import mockData from './mock_data'


const factory = () => {
    return shallowMount(PostView, {
        data() {
            return {post: mockData.posts[0]}
        },
        // mocks: {
        //     'post-card': function() {},
        // },
    });
}


describe('PostView render', () => {
    let wrapper;
    afterEach(() => {
        wrapper.destroy();
    });
    it('Render default', () => {
        wrapper = factory();
        console.log(wrapper.vm.$data);
        // wrapper.setData({post: mockData.posts[0]});
        expect(wrapper.html()).toMatchSnapshot();
    });
})

describe('PostView Events', () => {
    // let wrapper;
    // beforeEach(() => {
    //     wrapper = wrapper = factory({});
    // });
    // afterEach(() => {
    //     wrapper.destroy();
    // });
    // it('Action comment', () => {
    //     wrapper.findComponent(PostBody).vm.$emit('action-comment');
    //     expect(wrapper.emitted('action-post').length).toBe(1);
    // });
})