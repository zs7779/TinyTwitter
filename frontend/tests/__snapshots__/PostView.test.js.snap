// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostView render Render default 1`] = `
"<div>
  <post-card-stub verbose=\\"true\\" post=\\"[object Object]\\"></post-card-stub>
</div>"
`;
